### ISSUE TITLE
Request for `PIPENV_ONLY_USE_PYENV` environment variable.

### ISSUE THREAD

**Issue by BrennanGit** [pypa/pipenv#3855]
Be sure to check the existing issues (both open and closed!), and make sure you are running the latest version of Pipenv.

Check the [diagnose documentation](https://docs.pipenv.org/diagnose/) for common issues and the [PEEP list](https://github.com/pypa/pipenv/blob/master/peeps/) before posting! We may close your issue if it is very similar to one of them. Please be considerate and follow the PEEP process, or be on your way.

Make sure to mention your debugging experience if the documented solution failed.

### Additional context

Add any other context or screenshots about the feature request here. It may be a good idea to mention that platform and Python version you are on.

-------------------------------------------------------------------------------

Please run `$ pipenv --support`, and paste the results here. Don't put backticks (`` ` ``) around it! The output already contains Markdown formatting.

If you're on macOS, run the following:

    $ pipenv --support | pbcopy

If you're on Windows, run the following:

    > pipenv --support | clip

If you're on Linux, run the following:

    $ pipenv --support | xclip

**Comment by frostming** [comment]
tldr; set `PIPENV_PYTHON=$PYENV_ROOT/shims/pytho` in your `.bashrc` to force pipenv to use pyenv python.

Tell me if it solves your problem.

**Comment by BrennanGit** [comment]
Thanks for the massively quick reply.

> tldr; set `PIPENV_PYTHON=$PYENV_ROOT/shims/python` in your `.bashrc` to force pipenv to use pyenv python.
> 
> Tell me if it solves your problem.

Although this can be used to force pipenv to use a specific Python install, it does not really solve my problem. I'd still like the functionality provided where pyenv is used to install a python version that is not available, I'd just like a way to restrict which pythons could be considered usable by pipenv.

On a development machine, this would be a good solution (kind of equivalent to calling with `--python /path/to/python` I think?) but this is for a CI system and the goal is to allow developers to use whichever python version they require by specifying it in their Pipfile with the agent installing it with pyenv if it needs to.

**Comment by AlJohri** [comment]
@BrennanGit correct me if I'm wrong, but specifying `PIPENV_PYTHON=$PYENV_ROOT/shims/python`  should technically allow for any pyenv installed python, depending on which is currently active, right? it's basically `PIPENV_ONLY_USE_PYENV`?

**Comment by BrennanGit** [comment]
Yes this would allow you to force pipenv to use the currently active python from pyenv though this is not the behaviour I'm after. I want to be able to use pipenv+pyenv in it's normal (extremely useful) way but without worrying about other python installations (sometimes with incompatible compilation options) cluttering the system.

Here's an example:
- A user requests python 3.7 in their Pipfile and sends it off to the CI system.
- The CI machine has python 3.7 installed via brew (with broken Tkinter on Linux) and other lesser pythons installed with pyenv. We could uninstall this python but it is a dependency for many other packages.

What happens now:
- pipenv selects the broken python
- matplotlib fails

With `PIPENV_ONLY_USE_PYENV`:
- pipenv looks for python 3.7 but ignores any outside `PYENV_ROOT`
- pipenv gets pyenv to compile the right python
- we get a working version, compiled from source in a predictable, deterministic way

It would be very nice to not have to worry about all the weird system/package manager pythons that may be installed as dependencies from other packages. It just gives us a way to ensure pipenv reliably does the same thing which is very important for continuous integration.

**Comment by BrennanGit** [comment]
Hi guys, just hoping maybe you could change the label from "question" to "enhancement" so this doesn't get lost. Once I have a little more time, I'd be happy to implement this myself and do a PR.

I'd just like to check that this is functionality that is understood and wanted by anyone other than myself.

**Comment by atomic-mattie** [comment]
@BrennanGit I'm interested in what I think this functionality is just for development because:

- my `Pipfile` specifies `python_version = 3.7`
- `PIPENV_PYTHON=$HOME/.pyenv/shims/python` uses the currently-selected version (`system`), which isn't compatible
- I'd rather have pipenv search through my pyenv and find that I've installed 3.7.5
- when I upgrade my OS, any virtualenvs based on a system Python break, in my experience—pyenv Pythons are stable

**Comment by BrennanGit** [comment]
Hey @atomic-mattie, it does sound like your use case matches mine. The ability to restrict pipenv to use only pyenv-installed pythons would definitely help reproducability.

In the end we created a pipeline function in Jenkins that sets up a pipenv like this:
```
> parse python version from Pipfile
> pyenv install --list
> find best matching python with semver (same rules as pipenv uses)
> pyenv install -s $MATCHING
> pipenv install --python $(pyenv root)/versions/$MATCHING/bin/python" 
```
It's quite a rouandabout way of doing it but it forces our builds to be more repeatable, even across new machines and os updates.

**Comment by matteius** [comment]
I'd prefer if the environment variable were called `PIPENV_PYENV_ONLY` or something less verbose than `PIPENV_ONLY_USE_PYENV`.  Also what if someone wants to only use asdf (I've never used this)?   Maybe something more thoughtful to different modes for searching for Python, since there are still a number of issue reports about python finding logic floating around.  Just some thoughts as I comb through the backlog again.

**Comment by kalebmckale** [comment]
Perhaps I'm misunderstanding the issue, but when I set `pyenv local` versions, then `pipenv` will only use the installation of Python by `pyenv` with the version specified. When that version is not installed or not set by `pyenv local`, it asks me if I want `pyenv` to install that version.

**Comment by matteius** [comment]
@kalebmckale I think the original request is to exclude anything other that pyenv from the pythonfinder when a specific (new feature) env var gets set.

**Comment by kalebmckale** [comment]
Perhaps that's why I never had this issue. I am quite limited in the cases I've come across.
